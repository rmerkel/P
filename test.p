{ Calculate 11 (0..10) factorials
{	 n		    n!	}
{  ---  ---------	}
{	 0		    0	}
{	 1		    1	}
{	 2		    2	}
{	 3		    6	}
{	...				}
{	 9	  362,880	}
{	10	3,628,800	}
const nFacts = 10;
var n, f;	
begin
   n = 0;
   f = 1;
   while n < nFacts do begin
      n = n + 1;
      f = f * n
   end
end.

