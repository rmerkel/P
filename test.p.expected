# test.p, 2: { Calculate 11 (0..10) factorials
# test.p, 3: {	 n		    n!	}
# test.p, 4: {  ---  ---------	}
# test.p, 5: {	 0		    0	}
# test.p, 6: {	 1		    1	}
# test.p, 7: {	 2		    2	}
# test.p, 8: {	 3		    6	}
# test.p, 9: {	...				}
# test.p, 10: {	 9	  362,880	}
# test.p, 11: {	10	3,628,800	}
# test.p, 12: var n, f;
    0: call 0, 2
    1: halt
# test.p, 13: begin
    2: enter 2
# test.p, 14:    n = 0;
    3: pushConst 0
    4: pushVar 0, 4
    5: assign
# test.p, 15:    f = 1;
    6: pushConst 1
    7: pushVar 0, 5
    8: assign
# test.p, 16:    while n < 10 do begin
    9: pushVar 0, 4
   10: eval
   11: pushConst 10
   12: lt
   13: jneq 28
# test.p, 17:       n = n + 1;
   14: pushVar 0, 4
   15: eval
   16: pushConst 1
   17: addi
   18: pushVar 0, 4
   19: assign
# test.p, 18:       f = f * n
   20: pushVar 0, 5
   21: eval
# test.p, 19:    end
   22: pushVar 0, 4
   23: eval
   24: muli
   25: pushVar 0, 5
   26: assign
# test.p, 20: end.
   27: jump 9
   28: ret
#test.p, 20: 

        8:          0
        9:          1
        8:          1
        9:          1
        8:          2
        9:          2
        8:          3
        9:          6
        8:          4
        9:         24
        8:          5
        9:        120
        8:          6
        9:        720
        8:          7
        9:       5040
        8:          8
        9:      40320
        8:          9
        9:     362880
        8:         10
        9:    3628800
